mlrun/runtimes/base.py

```diff
@@ -1148,42 +1271,58 @@ def _ensure_runtime_resource_run_logs_collected(
                 store_log = True
 
         if store_log:
-            logger.info("Storing runtime resource log before deletion")
-            logs_from_k8s, _ = crud.Logs.get_log(
+            _, logs_from_k8s = crud.Logs.get_log(
                 db_session, project, uid, source=LogSources.K8S
             )
-            crud.Logs.store_log(logs_from_k8s, project, uid, append=False)
+            if logs_from_k8s:
+                logger.info("Storing finished run logs", project=project, uid=uid)
+                crud.Logs.store_log(logs_from_k8s, project, uid, append=False)
 
     @staticmethod
-    def _ensure_runtime_resource_run_status_updated(
+    def _ensure_run_status_updated(
         db: DBInterface,
         db_session: Session,
         project: str,
         uid: str,
         desired_run_state: str,
-    ):
-        run = db.read_run(db_session, uid, project)
-
-        current_run_state = run.get("status", {}).get("state")
-        logger.debug(
-            "Checking whether need to update run status",
-            desired_run_state=desired_run_state,
-            current_run_state=current_run_state,
-        )
+    ) -> str:
+        run = {}
         update_run = True
-        if current_run_state:
-            if current_run_state == desired_run_state:
-                update_run = False
-            # if the current run state is stable and different then the desired - don't touch
-            if current_run_state in RunStates.stable_states():
-                update_run = False
+        updated_run_state = desired_run_state
+        try:
+            run = db.read_run(db_session, uid, project)
+        except mlrun.errors.MLRunNotFoundError:
+            logger.warning(
+                "Run not found. A new run with status only will be created",
+                project=project,
+                uid=uid,
+                desired_run_state=desired_run_state,
+            )
+        else:
+            current_run_state = run.get("status", {}).get("state")
+            if current_run_state:
+                if current_run_state == desired_run_state:
+                    update_run = False
+                # if the current run state is stable and different then the desired - don't touch
+                if current_run_state in RunStates.stable_states():
```

Comment:
The comment repeats what the code does not EXPLAINS why.
This seems, at least to the uninformed reader, wrong - the `desired_run_state` is actually extracted from the actual k8s object (crd / pod) - so this is the "real" or "current" state.
The `current_run_state` is coming from mlrun's persistency, and is expected to be out of sync.
We're allegedly here updating mlrun's db state as per the real status. So why does mlrun's stored state trumps the real one?
The method's name says it "ensures" it's updated, here you choose to not update it.

Reply:
As I explained in the PR description, beside of this logic, the SDK running inside the Run pod itself (wrapping the user's code) also updates the Run's state.
So the scenario I was thinking about when doing this is:
The user's code finishes successfully, therefore the SDK updating the Run's state to `Completed` now for some reason whether because some bug in the wrapping SDK code, or I don't know, suddenly OOM kill to the Pod, the Pod got into Failed state.
So although in this scenario the user's code executed successfully we want to mark the Run as failed ? 
I'm not sure, it's a weird state, both decisions seems legit, anyways we must identify this state and log it.
what do you think ?

Reply:
i think a completed job/pod cannot fail after it's completed - completed is a state k8s puts after the pod's containers were terminated. once a runtime resource is marked completed, that's supposed to be it. so can't OOM kill a non running container

Reply:
As I wrote above, the scenario is that the user's code finishs running, then the SDK code (in the pod) updates the **Run's state** to completed (here the Pod is still running), now, if an exception will be raised (because some bug in the wrapping SDK or something), the Pod state will be changed to failed (from running!), and the API monitoring the Pod's state, will change the Run's state from one terminal state - completed - to another - error

Reply:
Talked offline, when the SDK will have an error we don't want to hide that so we're ok with the Run's state going from completed to failed.
We do identified that a race between SDK and server might occur so the Run's state will jitter: completed -> running -> completed, in the case where the API is monitoring the pod which is still running (cause SDK didn't finish yet) but the SDK already updated the Run's state to completed 
Will add a debounce logic for that